src/OpenTelemetry/Metrics/HistogramBuckets.cs

```diff
@@ -24,21 +26,56 @@ namespace OpenTelemetry.Metrics
     // Note: Does not implement IEnumerable<> to prevent accidental boxing.
     public class HistogramBuckets
     {
+        internal const int DefaultBoundaryCountForBinarySearch = 50;
+
         internal readonly double[] ExplicitBounds;
 
         internal readonly long[] RunningBucketCounts;
-
         internal readonly long[] SnapshotBucketCounts;
 
         internal double RunningSum;
-
         internal double SnapshotSum;
 
+        internal double RunningMin = double.PositiveInfinity;
+        internal double SnapshotMin;
+
+        internal double RunningMax = double.NegativeInfinity;
+        internal double SnapshotMax;
+
         internal int IsCriticalSectionOccupied = 0;
 
+        private readonly BucketLookupNode bucketLookupTreeRoot;
+
+        private readonly Func<double, int> findHistogramBucketIndex;
+
         internal HistogramBuckets(double[] explicitBounds)
         {
             this.ExplicitBounds = explicitBounds;
+            this.findHistogramBucketIndex = this.FindBucketIndexLinear;
+            if (explicitBounds != null && explicitBounds.Length >= DefaultBoundaryCountForBinarySearch)
+            {
+                this.bucketLookupTreeRoot = ConstructBalancedBST(explicitBounds, 0, explicitBounds.Length);
+                this.findHistogramBucketIndex = this.FindBucketIndexBinary;
+
+                static BucketLookupNode ConstructBalancedBST(double[] values, int min, int max)
```

Comment:
this does not seem related to adding MinMax to histograms.

Reply:
Ya, that's from the binary search PR I mentioned in the description that this branch was created from as it touches similar code.